hashicorp/dev-portal
# feat: allow toast component to work like dialog

- feat: implement toast skateboard using react-components
- feat: localize react-components, prep to rebuild for dev-dot UI
- fix: tweak toast positioning
- feat: stub in custom component
- feat: implement basic toast styles
- feat: add opt-in toast example to swingset
- chore: clean up CSS cruft
- docs: add comment on close-button
- fix: mistyped token
- chore: revert unnecessary dep bump
- fix: use alpha token for close-button background
- docs: add docs on ToastDisplayProps
- docs: add description on Toaster component
- docs: add description to toast component
- chore: split out toast-on-copy function
- chore: rename for consistency
- feat: allow toasts to not auto-dismiss
- fix: use duration Infinity, not loading toast
- feat: WIP implement interactive toast with dialog role
- fix: remove cruft test, now on separate page
- feat: add use-refocus hook, get things kinda working
- feat: use code-block copyButtonRef
- feat: support links and actions in interactive toast dialogs
## ✅ "Ready for Review" checklist


---

## 🔗 Relevant links



- [Preview - `/waypoint/tutorials?optInFrom=learn`][preview-tutorials-waypoitn] 🔎
- [Preview - `/vault/tutorials?optInFrom=learn`][preview-tutorials-vault] 🔎
- [Preview - `/swingset/components/toast`][preview-swingset] 🔎
- [Asana task][task] 🎟️

## 🗒️ What

Implements `components/toast`, and uses it to confirm success when an install command is copied on any `/<product>/downloads` view.

## 🤷 Why

- Our opt-in flow includes the use of toasts, so we need them there
- We want to more prominently highlight when install commands are copied, and we want to do this with a toast component

## 🛠️ How

- Implements toast components with [react-hot-toast]
    - Implementation is very similar to [@hashicorp/react-toast], and also draws inspiration from toasts in [hashicorp/learn], both of which also use [react-hot-toast]
    - Aims to match [HDS Toast component][hds-toast] 
- Pulls in a `code-block` update that implements an `onCopyCallback` prop
    - https://github.com/hashicorp/react-components/pull/591

## 📸 Design Screenshots

<img width="1440" alt="toast-stickersheet" src="https://user-images.githubusercontent.com/4624598/169442929-5551c3b7-1109-4929-8cf9-e7021bee7d0f.png">

## 🧪 Testing

- Visit an install page, such as [/waypoint/downloads][preview-downloads]
    - Navigate to an operating system tab that has code blocks for package manager install commands
    - When the `Copy` button in such a code block is activated, you should see both the state change to `Copied` within `code-block`, as well as a separate "Copied install command" toast pop up in the lower right of the screen.
- Visit both [/waypoint/downloads][preview-downloads] and [/vault/downloads][preview-downloads-vault], and confirm the above behaviour
- Visit [the Swingset page for Toast][preview-swingset] and confirm the component looks to spec
    - Some small elements, eg the `×`-close-button, were hard to track down in spec, so I matched what I could see at the [HDS Toast Component page][hds-toast]

![2022-05-19-download-toast-on-copy-demo](https://user-images.githubusercontent.com/4624598/169443108-b82c639c-1484-41fe-9c67-87c955f030a4.gif)

## 💭 Anything else?

- Intent is to build in this PR what we need for toasts for June 20
    - Apart from the downloads view `code-block` toast in this PR, I think this is only opt-in toasts
- Intent is _not_  to fully match the [HDS Toast Component][hds-toast] spec, _yet_
    - For example, arbitrary content such as italics, bold, and inline links are not _yet_ supported in the `description` prop. Intent is to implement that functionality later, if we need it. (May be sooner rather than later, I did see some design iterations that use inline links within the opt-in toasts, but I believe those are not the most current).

[preview-tutorials-waypoint]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/tutorials?optInFrom=learn
[preview-tutorials-waypoint]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/tutorials?optInFrom=learn
[preview-downloads]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/waypoint/downloads
[preview-downloads-vault]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/vault/downloads
[preview-swingset]: https://dev-portal-git-zstoast-components-hashicorp.vercel.app/swingset/components/toast
[react-hot-toast]: https://react-hot-toast.com
[@hashicorp/react-toast]: https://github.com/hashicorp/react-components/tree/main/packages/toast
[hashicorp/learn]: https://github.com/hashicorp/learn
[hds-toast]: https://design-system-components-hashicorp.vercel.app/components/toast
[task]: https://app.asana.com/0/1201903760348480/1202114097062999/f
